flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/window/TimeWindow.java

```diff
@@ -220,7 +220,9 @@ public TimeWindowSerializerSnapshot() {
      * @return window start
      */
     public static long getWindowStartWithOffset(long timestamp, long offset, long windowSize) {
-        return timestamp - (timestamp - offset + windowSize) % windowSize;
+        return timestamp
+                - (timestamp - offset) % windowSize
+                - (windowSize & (timestamp - offset) >> 63);
```

Comment:
`(windowSize & (timestamp - offset) >> 63` hard to understand, better to add more comments.